### ISSUE TITLE
question about custom objective

### ISSUE THREAD

**Issue by showkeyjar** [dmlc/xgboost#7325]
hi, I fig out a way to transform row weights data to objective function:
I add weights into label, and then in objective function, I subtract and got data row weights
because I can't find any function can read feature value in DMatrix, and xgb params has subsample.

but when I train model, the loss does not change, I want know what's wrong with my code:

my model is regression model

this is my objective function:

from scipy.misc import derivative

  def date_normal_obj(preds: np.ndarray, dtrain: xgb.DMatrix):
      fake_labels = dtrain.get_label()
      distance = np.floor(fake_labels / 1000).astype(np.int32)
      labels = ((fake_labels / 1000 - distance) * 1000).astype(np.float32)
      def dtn_loss(x):
          loss = np.sqrt(np.square(labels - x) / np.abs(distance) * max(distance))
          return loss
      grad = derivative(dtn_loss, preds, n=1, dx=1e-6)
      hess = derivative(dtn_loss, preds, n=2, dx=1e-6)
      return grad, hess

and this is my loss eval function:

  def date_normal_loss(predt: np.ndarray, dtrain: xgb.DMatrix):
      ''' Root mean squared error + date normal weight metric.'''
      fake_labels = dtrain.get_label()
      distance = np.floor(fake_labels / 1000).astype(np.int32)
      labels = ((fake_labels / 1000 - distance) * 1000).astype(np.float32)
      loss = np.sqrt(np.square(labels - predt) / np.abs(distance) * max(distance)).mean()
      return 'dtn-loss', loss

would you please help me find out what's my problem? thank you!

**Comment by trivialfis** [comment]
Subsampling is not related to DMatrix, you can use it without worrying lost data.

**Comment by trivialfis** [comment]
Is your loss differentiable?

**Comment by showkeyjar** [comment]
I just add mean() function in eval loss function, if not, train model will throw errors.

**Comment by trivialfis** [comment]
My question is whether your objective function is https://en.m.wikipedia.org/wiki/Differentiable_function , and can be approximated by https://en.m.wikipedia.org/wiki/Taylor_series. If not, can you prove a upper bound?

**Comment by showkeyjar** [comment]
I think if objective function is not differentiable, it will return same grad and hess result, but if I print grad and hess,it return different result every boost round. and I think the loss can not be same what ever objective function is differentiable.

**Comment by trivialfis** [comment]
Just use the normal weights in DMatrix.  As mentioned previously no need to worry about subsampling.  Gradient is calculated for all samples regardless of sampling.

**Comment by showkeyjar** [comment]
I mean the weight is my custom data row weight, not DMatrix weight.

**Comment by showkeyjar** [comment]
thanks for your patiently answer, I'll try weight param, by the way, if the weight.shape equals X rows, how to set feature weight?
